In [104]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings("ignore")

In [105]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(5)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                           1.08388                     7.60278   
ethereum                          0.22392                    10.38134   
tether                           -0.21173                     0.04935   
ripple                           -0.37819                    -0.60926   
bitcoin-cash                      2.90585                    17.09717   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                           6.57509                      7.67258   
ethereum                          4.80849                      0.13169   
tether                            0.00640                     -0.04237   
ripple                            2.24984                      0.23455   
bitcoin-cash                     14.75334                     15.74903   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                          -3.25185                      83.51840   
ethereum                        -12.88890                     186.77418   
tether                            0.28037                      -0.00542   
ripple                          -17.55245                      39.53888   
bitcoin-cash                    -13.71793                      21.66042   

              price_change_percentage_1y  
coin_id                                   
bitcoin                         37.51761  
ethereum                       101.96023  
tether                           0.01954  
ripple                         -16.60193  
bitcoin-cash                    14.49384

In [106]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [107]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [110]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_data_scaled = StandardScaler().fit_transform(df_market_data)

In [111]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(market_data_scaled, columns=['price_change_percentage_24h','price_change_percentage_7d','price_change_percentage_14d','price_change_percentage_30d','price_change_percentage_60d','price_change_percentage_200d','price_change_percentage_1y'])

# Copy the crypto names from the original DataFrame
df_market_data_scaled['coin_id'] = df_market_data.index

# Set the coin_id column as index
df_market_data_scaled = df_market_data_scaled.set_index('coin_id')

# Display the scaled DataFrame
df_market_data_scaled

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [114]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [115]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
for i in k:
    
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i, random_state=1)
    
# 2. Fit the model to the data using `df_market_data_scaled`
    k_model.fit(df_market_data_scaled)
    
# 3. Append the model.inertia_ to the inertia list
    inertia.append(k_model.inertia_)

In [116]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {'k': k, 'inertia': inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)
elbow_df

k     inertia
0   1  287.000000
1   2  195.820218
2   3  123.190482
3   4   79.022435
4   5   65.405923
5   6   52.933559
6   7   47.983124
7   8   37.288187
8   9   33.061685
9  10   28.779752

In [117]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
original_scaled_elbow_df = elbow_df.hvplot.line(
    x =  'k',
    y = 'inertia',
    label = 'Original Scaled Data'
)

original_scaled_elbow_df

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Either 3 or 4 could be good values for k in this case. I am choosing to select 4 for k because the curve has an additional small divot in the line before evening out, and I would like to see if a value of 4 can show us something additional in our data that a value of 3 may not show.

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [121]:
# Initialize the K-Means model using the best value for k
model = KMeans (n_clusters = 4, random_state=1)

In [122]:
# Fit the K-Means model using the scaled DataFrame
model.fit(df_market_data_scaled)

KMeans(n_clusters=4, random_state=1)

In [123]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
k_4 = model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
k_4

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 0, 3, 1, 1, 2, 1, 1, 1, 1])

In [124]:
# Create a copy of the scaled DataFrame
df_market_data_predictions = df_market_data_scaled.copy()

In [125]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_market_data_predictions['spread_4'] = k_4

# Display the copy of the scaled DataFrame
df_market_data_predictions

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [126]:
# Create a scatter plot using hvPlot by setting
original_scaled_plot = df_market_data_predictions.hvplot.scatter(
    
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
    x = 'price_change_percentage_24h',
    y = 'price_change_percentage_7d',
    
# Color the graph points with the labels found using K-Means and
    by = 'spread_4',
    
# add the crypto name in the `hover_cols` parameter to identify
 # the cryptocurrency represented by each data point.   
    hover_cols = ['coin_id'],
    label = 'Original Scaled Data'
)

original_scaled_plot

:NdOverlay   [spread_4]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [129]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [130]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
market_data_pca = pca.fit_transform(df_market_data_scaled)

# View the scaled PCA data
market_data_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [131]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
print(f"Explained Variance Ratio: {pca.explained_variance_ratio_}")
print(f"Sum of Explained Variance Ratio: {sum(pca.explained_variance_ratio_)}")

Explained Variance Ratio: [0.3719856  0.34700813 0.17603793]
Sum of Explained Variance Ratio: 0.8950316570309841


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The first principal component accounts for 37% of the data, the second component accounts for 35%, and the third accounts for 18%. The first two components account for a significant portion of the variability, whereas the third component contributes less influence. The combined sum of the explained variance ratio for all components is approximately 89.5%, which tells us that together they encompass a considerate amount of the variability in the dataset. This high total explained variance indicates that these three components are effective in representing the data.

In [133]:
# Create a new DataFrame with the PCA data.
df_market_data_pca = pd.DataFrame(market_data_pca, columns=['PC1', 'PC2', 'PC3'])

# Copy the crypto names from the original scaled DataFrame
df_market_data_pca['coin_id'] = df_market_data_scaled.index

# Set the coin_id column as index
df_market_data_pca = df_market_data_pca.set_index('coin_id')

# Display the scaled PCA DataFrame
df_market_data_pca

PC1       PC2       PC3
coin_id                                           
bitcoin              -0.600667  0.842760  0.461595
ethereum             -0.458261  0.458466  0.952877
tether               -0.433070 -0.168126 -0.641752
ripple               -0.471835 -0.222660 -0.479053
bitcoin-cash         -1.157800  2.041209  1.859715
binancecoin          -0.516534  1.388377  0.804071
chainlink            -0.450711  0.517699  2.846143
cardano              -0.345600  0.729439  1.478013
litecoin             -0.649468  0.432165  0.600303
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653
crypto-com-chain     -0.248198 -1.376252 -1.462026
usd-coin             -0.438408 -0.175337 -0.663388
eos                  -0.693425 -0.473815 -0.527597
monero                0.060499  2.909404  1.498571
tron                 -0.393352 -0.108192 -0.012756
tezos                -0.796176 -0.494409  1.082812
okb                   0.064075 -1.269825 -1.098829
stellar              -0.489015 -0.732719 -0.062543
cosmos               -0.306272  0.703415  1.714224
cdai                 -0.513528 -0.142802 -0.656566
neo                  -0.362120 -0.986914 -0.728752
wrapped-bitcoin      -0.604265  0.827398  0.439316
leo-token            -0.413296 -0.674115 -1.076628
huobi-token          -0.407483 -0.212507 -0.351426
nem                   0.608974  0.563532 -1.148742
binance-usd          -0.450211 -0.151019 -0.647401
iota                 -0.764665 -0.517886  0.204990
vechain              -0.556315 -1.938209 -1.261776
zcash                -0.425147  0.492976  1.058048
theta-token           2.676868 -0.013954 -1.965207
dash                 -0.613923 -0.479337  0.339565
ethereum-classic     -0.579924 -0.356334 -0.114942
ethlend               8.089018 -3.896891  2.301382
maker                -0.389045  0.165041  0.379414
havven                0.865762 -2.261882  0.275583
omisego               0.111675  0.428316 -1.205398
celsius-degree-token  4.792395  6.767679 -1.986985
ontology             -0.632355 -2.108117 -0.652227
ftx-token            -0.593142  0.021485  0.209911
true-usd             -0.458131 -0.135734 -0.635284
digibyte             -0.297910 -0.191126 -0.909602

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [136]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [137]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
for i in k:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model_pca = KMeans(n_clusters = i, random_state = 1)
# 2. Fit the model to the data using `df_market_data_pca`
    k_model_pca.fit(df_market_data_pca)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(k_model_pca.inertia_)

In [138]:
# Create a dictionary with the data to plot the Elbow curve
elbow_pca_data = {'k': k, 'inertia': inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_pca_df = pd.DataFrame(elbow_pca_data)
elbow_pca_df

k     inertia
0   1  256.874086
1   2  165.901994
2   3   93.774626
3   4   49.665497
4   5   38.352251
5   6   27.618972
6   7   21.134056
7   8   17.437664
8   9   13.742792
9  10   10.484890

In [139]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
pca_scaled_elbow_df = elbow_pca_df.hvplot.line(
    x = 'k',
    y = 'inertia',
    label = 'PCA Scaled Data'
)

pca_scaled_elbow_df

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No. Although 3 can be a good representative value of k, such as we saw with the original scaled data, I am choosing to go forward with 4 as the line follows an even and straight path with little curve as it continues.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [142]:
# Initialize the K-Means model using the best value for k
model_pca = KMeans(n_clusters = 4, random_state = 1)

In [143]:
# Fit the K-Means model using the PCA data
model_pca.fit(df_market_data_pca)

KMeans(n_clusters=4, random_state=1)

In [144]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pca_k_4 = model_pca.predict(df_market_data_pca)

# Print the resulting array of cluster values.
pca_k_4

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [145]:
# Create a copy of the scaled PCA DataFrame
df_market_data_pca_predictions = df_market_data_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_market_data_pca_predictions['spread_4'] = pca_k_4

# Display the copy of the scaled PCA DataFrame
df_market_data_pca_predictions

PC1       PC2       PC3  spread_4
coin_id                                                     
bitcoin              -0.600667  0.842760  0.461595         1
ethereum             -0.458261  0.458466  0.952877         1
tether               -0.433070 -0.168126 -0.641752         0
ripple               -0.471835 -0.222660 -0.479053         0
bitcoin-cash         -1.157800  2.041209  1.859715         1
binancecoin          -0.516534  1.388377  0.804071         1
chainlink            -0.450711  0.517699  2.846143         1
cardano              -0.345600  0.729439  1.478013         1
litecoin             -0.649468  0.432165  0.600303         1
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653         0
crypto-com-chain     -0.248198 -1.376252 -1.462026         0
usd-coin             -0.438408 -0.175337 -0.663388         0
eos                  -0.693425 -0.473815 -0.527597         0
monero                0.060499  2.909404  1.498571         1
tron                 -0.393352 -0.108192 -0.012756         0
tezos                -0.796176 -0.494409  1.082812         1
okb                   0.064075 -1.269825 -1.098829         0
stellar              -0.489015 -0.732719 -0.062543         0
cosmos               -0.306272  0.703415  1.714224         1
cdai                 -0.513528 -0.142802 -0.656566         0
neo                  -0.362120 -0.986914 -0.728752         0
wrapped-bitcoin      -0.604265  0.827398  0.439316         1
leo-token            -0.413296 -0.674115 -1.076628         0
huobi-token          -0.407483 -0.212507 -0.351426         0
nem                   0.608974  0.563532 -1.148742         0
binance-usd          -0.450211 -0.151019 -0.647401         0
iota                 -0.764665 -0.517886  0.204990         0
vechain              -0.556315 -1.938209 -1.261776         0
zcash                -0.425147  0.492976  1.058048         1
theta-token           2.676868 -0.013954 -1.965207         0
dash                 -0.613923 -0.479337  0.339565         0
ethereum-classic     -0.579924 -0.356334 -0.114942         0
ethlend               8.089018 -3.896891  2.301382         2
maker                -0.389045  0.165041  0.379414         1
havven                0.865762 -2.261882  0.275583         0
omisego               0.111675  0.428316 -1.205398         0
celsius-degree-token  4.792395  6.767679 -1.986985         3
ontology             -0.632355 -2.108117 -0.652227         0
ftx-token            -0.593142  0.021485  0.209911         0
true-usd             -0.458131 -0.135734 -0.635284         0
digibyte             -0.297910 -0.191126 -0.909602         0

In [146]:
# Create a scatter plot using hvPlot by setting
pca_scaled_plot = df_market_data_pca_predictions.hvplot.scatter(
# `x="PC1"` and `y="PC2"`.
    x = 'PC1',
    y = 'PC2',
# Color the graph points with the labels found using K-Means and
    by = 'spread_4',
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.   
    hover_cols = 'coin_id',
    label = 'PCA Scaled Data'
)

pca_scaled_plot

:NdOverlay   [spread_4]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [148]:
# Composite plot to contrast the Elbow curves
composite_elbow_plot = original_scaled_elbow_df + pca_scaled_elbow_df
composite_elbow_plot

:Layout
   .Curve.Original_Scaled_Data :Curve   [k]   (inertia)
   .Curve.PCA_Scaled_Data      :Curve   [k]   (inertia)

In [149]:
# Composite plot to contrast the clusters
composite_cluster_plot = original_scaled_plot + pca_scaled_plot
composite_cluster_plot

:Layout
   .NdOverlay.Original_Scaled_Data :NdOverlay   [spread_4]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.PCA_Scaled_Data      :NdOverlay   [spread_4]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** To ensure that clusters are separated in a way that is meaningful and accurately representative of the data while at the same time being manageable for analysis/interpretation, the choice in selecting a proper k value can be intimidating. Using fewer features can contribute to underfitting the data, to the point where the model is not detailed enough to show the structure of the data set. If clusters are too broad or lack detail, then they may not represent the data well. Lower k values can give higher variance, implying that data points are too similar to interpret well. Although interpretation may be easier, this can oversimply our model, with doesn't give us a lot of insight, or may generalize the data, leaving out the opportunity for complexities and potential findings. In our case here, a k value of 4 gave us insight into some data points that may qualify as outliers or may carry some complexity that the other data points do not follow. We can gain further insight by investigating what made these few data points different from our more clustered/grouped together data points. the data.